src/librustc_borrowck/borrowck/check_loans.rs

```diff
@@ -395,18 +395,33 @@ impl<'a, 'tcx> CheckLoanCtxt<'a, 'tcx> {
         assert!(self.bccx.region_scope_tree.scopes_intersect(old_loan.kill_scope,
                                                        new_loan.kill_scope));
 
-        self.report_error_if_loan_conflicts_with_restriction(
-            old_loan, new_loan, old_loan, new_loan) &&
-        self.report_error_if_loan_conflicts_with_restriction(
-            new_loan, old_loan, old_loan, new_loan)
+        let err_old_new = Result::err(self.report_error_if_loan_conflicts_with_restriction(
+            old_loan, new_loan, old_loan, new_loan
+        ));
+        let err_new_old = Result::err(self.report_error_if_loan_conflicts_with_restriction(
+            new_loan, old_loan, old_loan, new_loan
+        ));
+
+        if let Some(mut err_old) = err_old_new {
+            err_old.emit();
+            if let Some(mut err_new) = err_new_old {
+                err_new.cancel();
+            }
+        } else if let Some(mut err_new) = err_new_old {
+            err_new.emit();
+        } else {
+            return true;
+        }
```

Comment:
This block of code can be made slightly more idiomatic in the following way:

```rust
match (err_old_new, err_new_old) {
    (Some(mut err), None) | (None, Some(mut err)) => err.emit(),
    (Some(mut err_old), Some(mut err_new)) => {
        err_old.emit();
        err_new.cancel();
    }
    (None, None) => return true,
}
```

This way the compiler will always check that all paths have been taken care of, and you can handle cases that are mutually exclusive but that require the same code only once.